cips/cip-shwap_protocol.md

```diff
@@ -0,0 +1,378 @@
+compositions of Shwap with lower-level protocols Bitswap, KadDHT, or Shrex, which Shwap aims to describe.
+
+## Specification
+
+### Terms and Definitions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in BCP
+14 [RFC2119] [RFC8174] when, and only when, they appear in all
+capitals, as shown here.
+
+Commonly used terms in this document are described below.
+
+_**Shwap**_: The protocol described by this document. Shwap is a
+portmanteau name of words share and swap.
+
+_**[Share][shares]**_: The core data structure of DataSquare **"swapped"** between peers.
+
+_**[DataSquare][square]**_: The DA square format used by Celestia DA network.
+
+_**[DAH][dah]**_: The Data Availability Header with Row and Column commitments.
+
+_**[Namespace][ns]**_: The namespace grouping sets of shares.
+
+_**Peer**_: An entity that can participate in a Shwap protocol. There are three types of peers:
+client, server, and node.
+
+_**Client**_: The Peer that requests content by content identifies over Shwap.
+
+_**Server**_: The Peer that responds with content over Shwap.
+
+_**Node**_: The peer that is both the client and the server.
+
+_**Proof**_: A Merkle inclusion proof of the data in the DataSquare.
+
+### Message Framework
+
+This section defines Shwap's messaging framework. Every group of shares that needs to be exchanged over the network
+MUST define its [share identifier](#share-identifiers) and [share container](#share-containers), as well as, follow
+their described rules.
+
+#### Share Identifiers
+
+Identifiers MUST have a fixed size for their fields. Subsequently, protobuf SHOULD NOT be used for CID serialization due
+to varints and lack of fixed size arrays. Instead, identifiers use simple binary big endian serialization.
+
+Identifiers MAY embed each other to narrow down the scope of needed shares. For example, [SampleID](#sampleid) embeds
+[RowID](#rowid) as every sample lay on a particular row.
+
+#### Share Containers
+
+Share containers encapsulate a set of data shares with [DAH][dah] inclusion proof. Share containers are identified by
+[share identifiers](#share-identifiers).
+
+#### Versioning
+
+If a defined share container or identifier requires an incompatible change, the new message type MAY be introduced
+suffixed with a new major version starting from v1. E.g., if the Row message needs a revision, RowV1 is created.
+
+### Messages
+
+This section defines all the supported Shwap messages, including share identifiers and containers. All the new
+future messages should be described in the section.
+
+#### RowID
+
+RowID identifies the [Row shares container](#row-container) in a [DataSquare][square].
+
+RowID identifiers are formatted as shown below:
+
+```text
+RowID {
+    Height: u64;
+    RowIndex: u16;
```

Comment:
Not sure I understand the question.

Reply:
Not that I would be aware of. The u16 is chosen because it maxes at 1TB blocks, which is more than enough.

Reply:
cc @rootulp, do we have any row number limits defined in core?

Reply:
Yea! The max row number is equivalent to  2 * the max effective square size because the row number seems applicable to the extended data square. The max effective square size is currently 64 but it could be bumped to 128 via a governance proposal. 

Ref: https://gist.github.com/rootulp/bbf10f6e9cf114816aaa994eb64b63a4

Reply:
leaving this discussion open to migrate into the forum